### DISCUSSION TITLE
InvalidCastException between ViewModels when using A UserControl wrapping other controls in a parent UserControl.

### DISCUSSION THREAD

**Discussion by darcome** [AvaloniaUI/Avalonia#14239] [Q&A]
Hello everyone, 
it's a bit difficult for me explaining what is going on, therefore I'll use some screenshots and a repo, in case you want to see the full code.

First of al, here is the MyControl axaml that uses GroupBox and MyControlViewModel. On line 14 I use the NameFilter property of such viewmodel:

![image](https://github.com/AvaloniaUI/Avalonia/assets/18150093/726c3e14-4526-466c-a42b-d2f4e771e006)

As you can see the whole control is surronunded by GroupBox which is another UserControl that defines GroupBoxViewModel:

![image](https://github.com/AvaloniaUI/Avalonia/assets/18150093/3ce49e7a-215d-408a-8548-d9d93d5a5170)

Then the MyControl is used in MainWindow:

![image](https://github.com/AvaloniaUI/Avalonia/assets/18150093/ae37f5e4-83f3-4508-9141-b40daaa643b5)

As you can see, the previewer shows an InvalidCastException and the same happens when I start the app.

Here is the link to the repo: [https://github.com/darcome/TestRepo](https://github.com/darcome/TestRepo) in case you want to see the full code.

Can you please tell me what I am doing wrong and how to solve this problem?

Thanks in advance!

**Comment by thevortexcloud** [comment]
Does this actually display properly? It looks fairly odd to me. You are setting the content property which I think would be overriding your template? It really looks like you want a templated control here.

https://docs.avaloniaui.net/docs/concepts/templated-controls

On a side note, Avalonia has a similar control to your groupbox called `HeaderedContentControl`. I don't think the template has been released yet but you can copy it from here. 

https://github.com/AvaloniaUI/Avalonia/blob/1c792b294f82b26396e4f91032a2b535ea9c0669/src/Avalonia.Themes.Fluent/Controls/HeaderedContentControl.xaml

Also your view model is violating MVVM by directly having a control reference. The view model is not allowed to know about the view, however the view can know about the view model.

I would suggest rewriting this to follow the proper patterns. It would also make it much easier for me to follow.

**Comment by timunie** [comment]
@darcome check https://github.com/darcome/TestRepo/pull/1

**Comment by darcome** [comment]
Thank you everyone! I solved the problem by removing the viewmodel from the groupbox and making Text, Content and Header StyledProperties. Now everything works as expected and it doesn't violate MVVM pattern!
